runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -486,32 +486,70 @@ TR_J9ServerVM::classHasBeenReplaced(TR_OpaqueClassBlock *clazz)
    return ((classDepthAndFlags & J9AccClassHotSwappedOut) != 0);
    }
 
+bool
+TR_J9ServerVM::checkCHTableIfClassInfoExistsAndHasBeenExtended(TR_OpaqueClassBlock *clazz, bool &bClassHasBeenExtended)
+   {
+   JITServerPersistentCHTable *table = (JITServerPersistentCHTable*)(_compInfo->getPersistentInfo()->getPersistentCHTable());
+   TR_PersistentClassInfo *classInfo = table->findClassInfoAfterLocking(clazz, this);
+   if (classInfo)
+      {
+      bClassHasBeenExtended = classInfo->getFirstSubclass() ? true : false;
+      }
+   return (classInfo ? true : false);
+   }
+
 bool
 TR_J9ServerVM::classHasBeenExtended(TR_OpaqueClassBlock *clazz)
    {
-   if(!clazz)
+   if (!clazz)
       return false;
-   uintptr_t classDepthAndFlags = 0;
-   JITServer::ServerStream *stream = _compInfoPT->getMethodBeingCompiled()->_stream;
-   bool dataFromCache = JITServerHelpers::getAndCacheRAMClassInfo((J9Class *)clazz, _compInfoPT->getClientData(), stream, JITServerHelpers::CLASSINFO_CLASS_DEPTH_AND_FLAGS, (void *)&classDepthAndFlags);
 
-   if(dataFromCache)
+   bool bDataCached = false;
+   ClientSessionData *clientSessionData = _compInfoPT->getClientData();
       {
-      // Check if flag is set, since once set it can not be removed
-      if((classDepthAndFlags & J9AccClassHasBeenOverridden) != 0)
-         {
-         return true;
-         }
-      else
+      OMR::CriticalSection getRemoteROMClass(clientSessionData->getROMMapMonitor());
+      auto it = clientSessionData->getROMClassMap().find((J9Class*)clazz);
+      if (it != clientSessionData->getROMClassMap().end())
          {
-         //Flag not set, check with client as cache data may be outdated
-         stream->write(JITServer::MessageType::VM_classHasBeenExtended, clazz);
-         return std::get<0>(stream->read<bool>());
+         bDataCached = true;
+         if ((it->second._classDepthAndFlags & J9AccClassHasBeenOverridden) != 0)
+            {
+            return true;
+            }
+         else
+            {
+            bool bClassHasBeenExtended = false;
+            if (checkCHTableIfClassInfoExistsAndHasBeenExtended(clazz, bClassHasBeenExtended))
```

Comment:
> This acquires CHTable monitor and above we acquired the ROMMapMonitor. This imposes an order in which these monitors need to be acquired. 

> It may be easier to verify the code in JITServerPersistentCHTable.cpp. 

I was concerned about the potential deadlock if these two monitors are not taken in this order in other places. I didn't find a better solution to do it. I'll rethink about this part.